src/Microsoft.AspNetCore.Dispatcher/DispatcherLoggerExtensions.cs

```diff
@@ -0,0 +1,51 @@
+﻿using System;
+using System.Linq;
+using Microsoft.Extensions.Logging;
+
+namespace Microsoft.AspNetCore.Dispatcher.Internal
+{
+    internal static class DispatcherLoggerExtensions
+    {
+        // Too many matches
+        private static readonly Action<ILogger, string, Exception> _ambiguousEndpoints = LoggerMessage.Define<string>(
+            LogLevel.Error,
+            new EventId(1, "AmbiguousEndpoints"),
+            "Request matched multiple endpoints resulting in ambiguity. Matching endpoints: {AmbiguousEndpoints}");
+
+        // Unique match found
+        private static readonly Action<ILogger, Endpoint, Exception> _endpointMatched = LoggerMessage.Define<Endpoint>(
+            LogLevel.Information,
+            new EventId(1, "EndpointMatched"),
+            "Request matched endpoint: {endpoint}");
+
+        private static readonly Action<ILogger, string[], Exception> _noEndpointsMatched = LoggerMessage.Define<string[]>(
+            LogLevel.Debug,
+            new EventId(2, "NoEndpointsMatched"),
+            "No endpoints matched the current request. Dispatcher values: {DispatcherValueCollection}");
+
+        public static void AmbiguousEndpoints(this ILogger logger, string ambiguousEndpoints)
+        {
+            _ambiguousEndpoints(logger, ambiguousEndpoints, null);
+        }
+
+        public static void EndpointMatched(this ILogger logger, Endpoint endpoint)
+        {
+            _endpointMatched(logger, endpoint, null);
```

Comment:
@rynowak I wanted to use the Endpoint.DisplayName property here but I realized that it may sometimes not be set/be null.
What's a sensible logging message in that case?

Reply:
DisplayName is 100% the thing to use. If someone set it to null they can cry about it.